### ISSUE TITLE
How to use the results of (for example) gethistory?

### ISSUE THREAD

**Issue by FoxLisk** [fenhl/rust-racetime#3]
I am trying to write some code that does something like the following:

1. when a new race is registered, send a `gethistory` command[0]
2. wait on the result of that command
3. send a new message based on the results of that command

This seems to be very hard with the way this library currently works. The `new` handler must return before the `chat_history` handler is called. This is reasonable - the natural approach to this constraint is to run your race handler code in a separate tokio task and communicate with channels.

But I then run into another issue: if I have a separate task that's doing the handling, it can't actually send messages, because I only have access to the `&RaceContext<T>` during the execution of the callbacks.

now, the `&RaceContext<T>` has a perfectly cloneable `pub sender: Arc<Mutex<WsSink>>` but I would much rather use the convenience methods (`.send_message()`, etc).

Am I missing something? Is there a good solution here?

If I'm not, do you have any ideas for how to solve this? One that occurs to me is to split out the `send-messages-to-the-race-room` functionality into a separate struct, call it like `RaceSender`. a `RaceSender` can be a thin wrapper around the `Arc<Mutex<WsSink>>` with the convenience methods, and then it can be `#[derive(Clone)]` and consumers of the library can grab it as necessary.

Thanks!

-----

[0] The reason for this, in case it's relevant, is that I want the bot to be robust to restarts. On a restart it will rejoin all the rooms it was already handling and I'd rather it not repeat itself. Obviously I could handle this by persisting state somewhere, but it doesn't seem like that should be necessary. And in general, having a non-reactive handle to send messages to a race room seems like a reasonable ask.

**Comment by fenhl** [comment]
`RaceContext` implements `Clone`, which is cheap since internally it's just 3 `Arc`s. You can keep a `RaceContext` around, call its `data` method to check the current state of the room, and call the other methods to interact with it. (The `sender` field is really just an implementation detail and should probably be private.)

I'll have to improve the docs to explain this better, keeping this issue open for that purpose.

**Comment by FoxLisk** [comment]
Oh does it? Is that new? It didn't in the version I'm running, maybe I need to upgrade. that's an easy enough solution though, thank you!
